### PR TITLE
Improve ImmutableArrayBuilder<T> type in MVVM Toolkit generators

### PR DESCRIPTION
**Pull request by Sergio0694**
## Overview

This PR includes a couple tweaks to `ImmutableArrayBuilder<T>`:
- Drops the `ObjectPool<T>` use, and just uses `ArrayPool<T>`. This means we can now also benefit from the automatic trimming that `ArrayPool<T>` implements if there is high GC pressure. Also, we're no longer reinventing the wheel here.
- Make the type a `ref struct`, since we're never meant to pass this outside or box it.
- Add ref scope annotations where needed

## PR Checklist

- [X] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [X] Based off latest main branch of toolkit
- [X] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [X] Tested code with current [supported SDKs](../#supported)
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run _build/UpdateHeaders.bat_)
- [X] Contains **NO** breaking changes
- [X] Every new API (including internal ones) has full XML docs
- [X] Code follows all style conventions

### CODE REVIEW

File: src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs [hunk 1]
```diff
@@ -15,13 +17,8 @@ namespace CommunityToolkit.Mvvm.SourceGenerators.Helpers;
 /// A helper type to build sequences of values with pooled buffers.
 /// </summary>
 /// <typeparam name="T">The type of items to create sequences for.</typeparam>
-internal struct ImmutableArrayBuilder<T> : IDisposable
+internal ref struct ImmutableArrayBuilder<T>
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs:20]: Can't implement interfaces on `ref struct` types. But Roslyn has special support for them via duck-typing, so as long as they expose a `public void Dispose()` method, you can still use such types in `using` statements/expressions 🙂

File: src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs [hunk 2]
```diff
@@ -96,30 +94,25 @@ public int Count
...
     /// A class handling the actual buffer writing.
     /// </summary>
-    private sealed class Writer
+    private sealed class Writer : IDisposable
     {
         /// <summary>
         /// The underlying <typeparamref name="T"/> array.
         /// </summary>
-        private T[] array;
+        private T?[]? array;
```
Comment (Youssef1313) [inline src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs:115]: The array is assigned to non-null in constructor. Anything I'm missing?

Comment (Youssef1313) [reply src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs:115]: Oh I see, it's null-ed in Dispose.

**Review by Youssef1313 — APPROVED** [review]
LGTM

**Pull request merged**